clash-prelude/src/Clash/Explicit/Reset.hs

```diff
@@ -225,45 +229,164 @@ resetSynchronizer clk rst = rstOut
+-- power-on reset, or 'holdReset' if filtering is only needed on deassertion.
+--
+-- @unsafeResetGlitchFilter@ allows breaking the requirement of initial values,
+-- but by doing so it is possible that the design starts up with a period of up
+-- to /2 * glitchlessPeriod/ clock cycles where the reset output is unasserted
+-- (or longer in the case of glitches on the filtered reset input). This can
+-- cause a number of problems. The outputs\/tri-states of the design might
+-- output random things, including coherent but incorrect streams of data. This
+-- might have grave repercussions in the design's environment (sending network
+-- packets, overwriting non-volatile memory, in extreme cases destroying
+-- controlled equipment or causing harm to living beings, ...).
+--
+-- Without initial values, the synthesized result of @unsafeResetGlitchFilter@
+-- eventually correctly outputs a filtered version of the reset input. However,
+-- in simulation, it will indefinitely output an undefined value. This happens
+-- both in Clash simulation and in HDL simulation. Therefore, simulation should
+-- not include the @unsafeResetGlitchFilter@.
+--
+-- If @unsafeResetGlitchFilter@ is used in a domain with asynchronous resets
+-- ('Asynchronous'), @unsafeResetGlitchFilter@ will first synchronize the reset
+-- input with 'dualFlipFlopSynchronizer'.
+unsafeResetGlitchFilter
+  :: forall dom glitchlessPeriod
+   . ( HasCallStack
+     , KnownDomain dom
+     , 1 <= glitchlessPeriod
+     )
+  => SNat glitchlessPeriod
+  -- ^ Consider a reset signal to be properly asserted after having seen the
+  -- reset asserted for /glitchlessPeriod/ cycles straight.
+  -> Clock dom
+  -> Reset dom
+  -> Reset dom
+unsafeResetGlitchFilter glitchlessPeriod clk =
+  resetGlitchFilter# glitchlessPeriod reg dffSync
+ where
+  reg = delay clk enableGen
+  dffSync = dualFlipFlopSynchronizer clk clk noReset enableGen
+{-# INLINE unsafeResetGlitchFilter #-}
+
+-- | Filter glitches from reset signals by only triggering a reset after it has
+-- been asserted for /glitchlessPeriod/ cycles. Similarly, it will stay
+-- asserted until a /glitchlessPeriod/ number of deasserted cycles have been
+-- observed.
+--
+-- Compared to 'resetGlitchFilter', this function adds an additional power-on
+-- reset input. As soon as the power-on reset asserts, the reset output will
+-- assert, and after the power-on reset deasserts, the reset output will stay
+-- asserted for another /glitchlessPeriod/ clock cycles. This is identical
+-- behavior to 'holdReset' where it concerns the power-on reset, and differs
+-- from the filtered reset, which will only cause an assertion after
+-- /glitchlessPeriod/ cycles.
+--
+-- If @resetGlitchFilterWithReset@ is used in a domain with asynchronous resets
+-- ('Asynchronous'), @resetGlitchFilterWithReset@ will first synchronize the
+-- reset input with 'dualFlipFlopSynchronizer'.
+resetGlitchFilterWithReset
```

Comment:
I think I like `WithReset` better even though it's rather strange. Especially because our rules on letter case make it sound like cut pork (the _k_'s been cut :-P).

Reply:
Cool :)